### PR TITLE
Support multi-county audits

### PR DESCRIPTION
**Pull request by sduckett**
## Overview

This PR covers two closely related stories: _[Calculate appropriate diluted margin for all contests, including multi-county](https://www.pivotaltracker.com/story/show/159572709)_ and _[Select random samples from grouped ballot manifests, where necessary](https://www.pivotaltracker.com/story/show/159573106)_.

We are adding the ability to do audits of contests spanning one or more reporting counties (districts.) By looking at the set of districts that have reported sets of results for consistently named contests, we derive margins and vote totals for each contest. We look at an aggregate, ordered ballot manifest (or universe) based on the counties participating in each targeted contest.

## How it works

Once counties have uploaded their manifests and CVRs, we define an audit, standardize contest names so we can group contests into _ballot universes_, and select the contests to audit. Once an audit is launched and [round one started](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-65c08163a7431a6ee7c9a63a4e62a8afR206), the first step is to [group and count contests by name](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-65c08163a7431a6ee7c9a63a4e62a8afR153) to arrive at vote totals for each contest. A [`ContestResult` represents the results for a contest across one or more counties.](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-f19eead924393b1a951f40c3dc4e10f5R38). Once they're counted, we can [select random ballots for each targeted contest](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-65c08163a7431a6ee7c9a63a4e62a8afR216), and [pick the ballots for each county](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-65c08163a7431a6ee7c9a63a4e62a8afR235). Finally, we [setup comparison audits for each county](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-8e866f810560a91e0109e02b3cc943f2R337).

## Every contest is a multi-county contest

Rather than considering "county-wide" contests differently than "multi-county", differently than "statewide" contests, we view all of the contests grouped by name, so some contests will have one participating county, while others will have many. This makes the setup consistent: we don't have to consider special cases when initializing audits.

## Diluted margins

Nothing has changed about _how_ a diluted margin is calculated. The difference is that the number of ballots cast and minimum margin are now larger numbers because many counties are contributing ballots. 

We chose to [extract some math into a new `Audit` class](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-ecc8d79627d24b917beae17179a46f4eR1). Ideally, we'd remove all of the comparison audit math from the controller and give it test coverage. This is the first step in that direction.

## Ballot Universes

Once contests are grouped, we know which counties are participating. If we collect all `BallotManifestInfo` for these counties, ordered by `county_id`, we get a bigger ballot manifest view based on minimum `sequenceStart` and maximum `sequenceEnd` for each county. Each county's manifest offsets the next by some number. Now that we know the size of the ballot universe for this contest, we ask a `PRNG` for an audit sequence, which we then divide into a collection of county subsequences based on which manifest holds a given sample. Here's a short list of how that works.

- [Find the manifest entry for a random sample](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-68b624766742f8be3b009cfef603db9eR198).
- [Create a random list of numbers and divide them into the appropriate counties](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-68b624766742f8be3b009cfef603db9eR37)
- A `BallotManifestInfo` knows if [it is holding a sample](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-0225216a8f4be465a3f4384632a6dee6R152)

## Tests

If you place a test in the `integration` group, [it will be excluded by default](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-04c6e90faac2675aa89e2176d2eec7d8R93):

```java
public class IntegrationTest {
  @Test(groups = {"integration"})
  public void testSomethingIntegral() { ,,, }
```
See also: [test groups](http://testng.org/doc/documentation-main.html#test-groups)

## /etc

[BallotManifestDownloadByCounty](https://github.com/democracyworks/ColoradoRLA/pull/37/files#diff-e0a2fbc82ef3225ddabe65f8b01f400dR36) was incorrectly treating `county_id` as an Integer.

We've setup a path for integration tests playing nicely with TravisCI, but we still need to add the integration tests themselves. A minimal example of test setup for ballot selection exists, but needs a little more work before it is merged.

There's probably a little dead code we can find with a linter.

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/StartAuditRound.java [hunk 1]
```diff
@@ -114,33 +126,124 @@ public String endpointBody(final Request the_request,
...
+      .map(cr -> {cr.setAuditReason(targetedContestReasons(cta)
+                                    .getOrDefault(cr.getContestName(),
+                                                  AuditReason.OPPORTUNISTIC_BENEFITS));
+                  return cr; })
+      .map(Persistence::persist)
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * Select random ballots for each targeted contest. Returns
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/StartAuditRound.java]: 🎏 What does it return?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/ContestResult.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+  /**
+   * set the set of counties
+   */
+  public boolean addCounties(final Set<County> cs) {
+    return this.counties.addAll(cs);
+  }
+
+  /**
+   * set the set of counties
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/model/ContestResult.java]: 🎏 `s/counties/contests/`

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/StartAuditRound.java [hunk 3]
```diff
@@ -114,33 +126,124 @@ public String endpointBody(final Request the_request,
...
+      new ArrayList<Map<Long, List<Integer>>>();
+
+    for(final ContestResult contestResult: targetedContestResults) {
+      final BigDecimal optimistic =
+        Audit.optimistic(riskLimit, contestResult.getDilutedMargin());
+
+      Main.LOGGER.info("contest: " + contestResult.getContestName());
+      Main.LOGGER.info("riskLimit: " + riskLimit);
+      Main.LOGGER.info("dilutedMargin: " + contestResult.getDilutedMargin());
+      Main.LOGGER.info("optimistic: " + optimistic);
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/StartAuditRound.java]: collapse into single logger call

File: server/eclipse-project/src/main/java/us/freeandfair/corla/controller/BallotSelection.java [hunk 4]
```diff
@@ -27,8 +33,99 @@
...
+
+  /**
+   * @description add an element to a list, l, where l may be null.
+   **/
+  @SuppressWarnings({"PMD.AvoidReassigningParameters"}) // FIXME Don't reassign.
+  public static <T> List<T> addToList(List<T> l, final T t) {
+    l = (null == l) ? new ArrayList<T>() : l;
+    l.add(t);
+    return l;
+  }
```
[the remark this answers is not in the document]
Comment (sduckett) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/controller/BallotSelection.java]: It is used here: https://github.com/democracyworks/ColoradoRLA/pull/37/files/b385bd34204d9162ae34d9a5785cf63344fd9f2f#diff-68b624766742f8be3b009cfef603db9eR81

File: server/eclipse-project/src/main/java/us/freeandfair/corla/controller/ComparisonAuditController.java [hunk 5]
```diff
@@ -313,6 +312,9 @@ private ComparisonAuditController() {
     // round numbers are 1-based, not 0-based
     final Round round = the_cdb.rounds().get(the_round - 1);
 
+    Main.LOGGER.info("round.ballotSequence.size() = " + round.ballotSequence().size());
+    Main.LOGGER.info("round.ballotSequence() = " + round.ballotSequence());
```
[the remark this answers is not in the document]
Comment (sduckett) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/controller/ComparisonAuditController.java]: Yep, I like this flavor the best.

File: README.md [hunk 6]
```diff
@@ -73,6 +73,31 @@ docker-compose exec postgresql \
...
 ```
 
+Note that `coloradorla_postgresql_1` is the output of "Name" from
+
+```sh
+docker-compose ps postgresql
+```
+
+but should be consistent assuming you have checked out the code to a directory
+named the same as the repository.
```
[the remark this answers is not in the document]
Comment (sduckett) [reply README.md]: I probably overlooked it when resolving conflicts. 👍

**Comment by teaforthecat** [comment]
@nealmcb This was just meant to be ballot selection, could we address the discrepancies in another pr?

**Comment by nealmcb** [comment]
@teaforthecat if that's the way you'd like to parcel the work into PRs, that's fine.  I figured not always doing a uniform selection seems like a ballot selection issue, and so made sense here.  I.e. this isn't about which discrepancies trigger escalation, but about how selection for escalation happens.
Let me know if you want me to go ahead and approve this (preferably with a link to an issue for doing uniform selections).

**Comment by nealmcb** [comment]
Current work on this is happening on another branch [multi-county-discrepancies](https://github.com/democracyworks/ColoradoRLA/tree/multi-county-discrepancies) that forked from this branch (multi-county-selection).
Do we want to absorb the other commits from that branch into this one? Or perhaps there is another good approach?

**Comment by gws** [comment]
:wave: Is this PR and/or branch currently being used?

**Comment by sduckett** [comment]
No, we can close this one also.

**Comment by nealmcb** [comment]
I'm just leaving a trail of crumbs from this great description of the approach used in those branches that inherited from this one, the latest of which is #44.

**Pull request closed without merging**